src/crypto/tls/ech.go

```diff
@@ -0,0 +1,874 @@
+		if c.hrrTriggered && c.ech.st.Offered {
+			// Detect illegal second ClientHello.
+			c.sendAlert(alertIllegalParameter)
+			return nil, fmt.Errorf("ech: hrr: bypass after offer")
+		}
+
+		if config.echCanAccept() && hello.encryptedClientHelloOffered {
+			// If the "encrypted_client_hello" extension was set, then presume
+			// it was covertext.
+			c.ech.st.Grease = true
+		}
+
+		// Bypass ECH.
+		return hello, nil
+	}
+
+	if c.hrrTriggered && !c.ech.st.Offered {
+		// Detect illegal second ClientHello.
+		c.sendAlert(alertIllegalParameter)
+		return nil, fmt.Errorf("ech: hrr: offer after bypass")
+	}
+	c.ech.st.Offered = true
+
+	// Parse the payload of the ECH extension.
+	rawEch := hello.encryptedClientHello
+	ech, err := echUnmarshalClient(rawEch)
+	if err != nil {
+		c.sendAlert(alertIllegalParameter)
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	// TODO(cjpatton): DON'T MERGE: Ensure this is specified.
+	if c.hrrTriggered &&
+		(ech.handle.suite != c.ech.hrrHandle.suite ||
+			!bytes.Equal(ech.handle.configId, c.ech.hrrHandle.configId)) {
+		// Detect illegal second ClientHello.
+		c.sendAlert(alertIllegalParameter)
+		return nil, fmt.Errorf("ech: hrr: handle mismatch")
+	}
+	c.ech.hrrHandle = &ech.handle
+
+	// Ask the ECH provider for the decryption context. Note that c.ech.hrrPsk
+	// is initially nil, meaning no PSK is used for encrypting the first
+	// ClientHelloInner.
+	res, err := echProvider.GetServerContext(ech.handle.marshal(), c.ech.hrrPsk, extensionECH)
+	if err != nil {
+		// This condition indicates the connection must be aborted.
+		c.sendAlert(alert(res.Alert))
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	if res.Rejected {
+		if c.hrrTriggered && c.ech.st.Accepted {
+			// ECH was accepted prior to HRR then rejected after. Because the
+			// context handle is the same in the first ClientHello as it is in
+			// the second, this indicates a server failure, likely due to the
+			// ECH key being rotated. Attempting to recover here may trigger yet
+			// another HRR (this time from the client-facing server), so hard
+			// fail.
+			//
+			// TODO(cjpatton): Ensure this is specified.
+			c.sendAlert(alertHandshakeFailure)
+			return hello, fmt.Errorf("ech: hrr: reject after accept")
+		}
+		// Reject ECH.
+		c.ech.retryConfigs = res.RetryConfigs
+		return hello, nil
+	}
+
+	ctx, err := echUnmarshalServerContext(res.Context)
+	if err != nil {
+		c.sendAlert(alertInternalError)
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	rawCompressedHelloInner, err := ctx.decrypt(ech.payload)
+	if err != nil {
+		c.sendAlert(alertDecryptError)
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	rawHelloInner, innerDigest, ok := echUnincorporateOuterExtensions(rawCompressedHelloInner, hello.marshal())
+	if !ok {
+		c.sendAlert(alertIllegalParameter)
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	if innerDigest != nil {
+		if subtle.ConstantTimeCompare(innerDigest, ctx.innerDigest(rawHelloInner)) != 1 {
```

Comment:
Is constant-time comparison necessary here? If so, can we leave a comment saying why?

Reply:
My thinking is that timing info leakage could be a concern if the encryption scheme were malleable. It's not, so this may be overly conservative.

Reply:
Yeah, given that we're using an AEAD, I don't think this is necessary. But it also doesn't really hurt.

Reply:
I'll leave it as-is and leave a note.